File: sdk/python/testvectors/ReceiptFactory.py
from symbolchain import sc
from symbolchain.RuleBasedTransactionFactory import RuleBasedTransactionFactory
from symbolchain.symbol.Network import Address


class ReceiptFactory:
	"""Factory for creating Symbol receipts."""

	def __init__(self, type_rule_overrides=None):
		"""Creates a factory."""
		self.factory = self._build_rules(type_rule_overrides)

	def create(self, receipt_descriptor):
		"""Creates a receipt from a receipt_descriptor."""
		return self.factory.create_from_factory(sc.ReceiptFactory.create_by_name, receipt_descriptor), receipt_descriptor

	@staticmethod
	def _build_rules(type_rule_overrides):
		factory = RuleBasedTransactionFactory(sc, None, type_rule_overrides)
		factory.autodetect()

		factory.add_struct_parser('Mosaic')

		sdk_type_mapping = {
			'Address': Address,
		}
		for name, typename in sdk_type_mapping.items():
			factory.add_pod_parser(name, typename)

		return factory

File: sdk/javascript/vectors/ReceiptFactory.js
import RuleBasedTransactionFactory from '../src/RuleBasedTransactionFactory.js';
import { Address } from '../src/symbol/Network.js';
import * as sc from '../src/symbol/models.js';

/**
 * Factory for creating Symbol receipts.
 */
export default class ReceiptFactory {
	/**
	 * Creates a factory.
	 * @param {Map<string, function>|undefined} typeRuleOverrides Type rule overrides.
	 */
	constructor(typeRuleOverrides = undefined) {
		this.factory = ReceiptFactory._buildRules(typeRuleOverrides); // eslint-disable-line no-underscore-dangle
	}

	/**
	 * Creates a receipt from a receipt descriptor.
	 * @param {object} receiptDescriptor Receipt descriptor.
	 * @returns {object} Newly created receipt.
	 */
	create(receiptDescriptor) {
		return this.factory.createFromFactory(sc.ReceiptFactory.createByName, receiptDescriptor);
	}

	static _buildRules(typeRuleOverrides) {
		const factory = new RuleBasedTransactionFactory(sc, undefined, typeRuleOverrides);
		factory.autodetect();

		factory.addStructParser('Mosaic');

		const sdkTypeMapping = {
			Address
		};
		Object.keys(sdkTypeMapping).forEach(name => { factory.addPodParser(name, sdkTypeMapping[name]); });

		return factory;
	}
}
